src/mapwindow.jl

```diff
@@ -116,6 +123,89 @@ function _mapwindow{T,N}(f,
     out
 end
 
+
+
+# This is a implementation of Fast 2D median filter
+# http://ieeexplore.ieee.org/document/1163188/
+function median_filter{T,N}(f,
+                         img::AbstractArray{T,N},
+                         window::Indices{N},
+                         border::BorderSpecAny,
+                         shape=default_shape(f);
+                         callmode=:copy!)
+    inds = indices(img)
+    inner = _interior(inds, window)
+    if callmode == :copy!
+        buf = Array{T}(map(length, window))
+        bufrs = shape(buf)
+        Rbuf = CartesianRange(size(buf))
+        offset = CartesianIndex(map(w->first(w)-1, window))
+        # To allocate the output, we have to evaluate f once
+        Rinner = CartesianRange(inner)
+        # Initialise the mode to zero and histogram consisting of 255 bins to zeros
+        mode = 0
+        m_histogram=zeros(Int64,(256,))
+        if !isempty(Rinner)
+            out = similar(img, typeof(f(bufrs,m_histogram,mode,window)))
+            Rwin = CartesianRange(map(+, window, first(Rinner).I))
+            copy!(buf, Rbuf, img, Rwin)
+            prev_mode=0
+            prev_col=1            
+            for I in Rinner
+                curr_col=I.I[2]
+                if(column_change(curr_col,prev_col))
+                    m_histogram=zeros(Int64,(256,))
+                    prev_mode=0
+                end
+                prev_col=curr_col
+                
+                Rwin = CartesianRange(map(+, window, I.I))                                
+                copy!(buf, Rbuf, img, Rwin)
```

Comment:
Maybe I'm missing something, but won't this alone keep your method `O(n^2)` in 2d? You're copying all the values even if you only "use" some of them.